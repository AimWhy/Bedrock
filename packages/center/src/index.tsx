import { mergeBreakpoints } from "@bedrock-layout/spacing-constants";
import PropTypes from "prop-types";
import styled from "styled-components";

export interface CenterProps {
  maxWidth?: number | string;
  centerText?: boolean;
  centerChildren?: boolean;
}

function getSafeMaxWidth(
  breakPoints: Record<string, unknown>,
  maxWidth?: number | string
) {
  if (typeof maxWidth === "string") {
    return maxWidth;
  }

  return typeof maxWidth === "number"
    ? `${maxWidth}px`
    : mergeBreakpoints(breakPoints).medium + "px";
}

export const Center = styled.div.attrs<CenterProps>(() => {
  return {
    "data-bedrock-layout-center": "",
  };
})<CenterProps>`
  @property --maxWidth {
    syntax: "<length>";
    inherits: false;
    initial-value: 1023px;
  }
  --maxWidth: ${({ maxWidth, theme: { breakPoints } }) =>
    getSafeMaxWidth(breakPoints, maxWidth)};

  box-sizing: content-box;

  margin-inline-start: auto;
  margin-inline-end: auto;
  margin-inline: auto;

  max-inline-size: 1023px;

  @supports (
    max-inline-size:
      ${({ maxWidth, theme: { breakPoints } }) =>
        getSafeMaxWidth(breakPoints, maxWidth)}
  ) {
    max-inline-size: var(--maxWidth, 1023px);
  }

  ${(props) => (props.centerText ? "text-align: center;" : "")}

  ${(props) =>
    props.centerChildren
      ? `
  display: flex;
  flex-direction: column;
  align-items: center;
  `
      : ""}
`;

Center.displayName = "Center";

Center.propTypes = {
  maxWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  centerText: PropTypes.bool,
};
