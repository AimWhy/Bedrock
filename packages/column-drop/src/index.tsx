import {
  CSSLength,
  Gutter,
  SizesOptions,
  getSafeGutter,
  getSizeValue,
  useTheme,
} from "@bedrock-layout/spacing-constants";
import { forwardRefWithAs } from "@bedrock-layout/type-utils";
import React, { CSSProperties } from "react";

type MinItemWidth = CSSLength | number | SizesOptions;

export interface ColumnDropProps {
  gutter?: Gutter;
  minItemWidth?: MinItemWidth;
  noStretchedColumns?: boolean;
}

function getSafeMinItemWidth(
  theme: {
    space?: { [key: string]: string };
    sizes?: { [key: string]: string };
  },
  minItemWidth?: MinItemWidth,
) {
  return getSizeValue(theme, minItemWidth);
}

export const ColumnDrop = forwardRefWithAs<"div", ColumnDropProps>(
  (
    { as, gutter, style, minItemWidth, noStretchedColumns = false, ...props },
    ref,
  ) => {
    const theme = useTheme();
    const maybeGutter = getSafeGutter(theme, gutter);

    const attributeValue =
      noStretchedColumns === true ? "no-stretched-columns" : "";

    const safeMinItemWidth = getSafeMinItemWidth(theme, minItemWidth);
    const safeStyle = style ?? {};

    const Component = as ?? "div";
    return (
      <Component
        ref={ref}
        data-bedrock-column-drop={attributeValue}
        style={
          {
            ...safeStyle,
            "--gutter": maybeGutter,
            "--minItemWidth": safeMinItemWidth,
          } as CSSProperties
        }
        {...props}
      />
    );
  },
);
ColumnDrop.displayName = "ColumnDrop";
