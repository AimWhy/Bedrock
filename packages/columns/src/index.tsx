import { Gutter, getSafeGutter } from "@bedrock-layout/spacing-constants";
import { Stack, StackProps } from "@bedrock-layout/stack";
import {
  PolymorphicComponentPropsWithRef,
  PolymorphicRef,
} from "@bedrock-layout/type-utils";
import { useContainerQuery } from "@bedrock-layout/use-container-query";
import React, { ElementType, forwardRef } from "react";
import styled from "styled-components";

interface ColumnsBaseProps {
  gutter?: Gutter;
  columns?: number;
  dense?: boolean;
  forwardedAs?: unknown;
}

const ColumnsBase = styled.div.attrs<ColumnsBaseProps>(
  ({ dense, theme, gutter, columns = 1, style }) => {
    const maybeGutter = getSafeGutter(theme, gutter);
    const safeColumns = columns > 0 ? columns : 1;
    return {
      "data-bedrock-columns": dense ? "dense" : "",
      style: { ...style, "--gutter": maybeGutter, "--columns": safeColumns },
    };
  }
)<ColumnsBaseProps>`
  @property --gutter {
    syntax: "<length-percentage>";
    inherits: false;
    initial-value: 0;
  }

  @property --columns {
    syntax: "<number>";
    inherits: true;
    initial-value: 1;
  }

  box-sizing: border-box;
  > * {
    margin: 0;
  }

  display: grid;
  grid-template-columns: repeat(var(--columns, 1), 1fr);
  gap: var(--gutter, 0px);
  grid-auto-flow: row ${({ dense = false }) => (dense === true ? "dense" : "")};
`;

interface ColumnsPropsBase extends StackProps, ColumnsBaseProps {
  switchAt?: number | string;
  children?: React.ReactNode;
}

export type ColumnsProps<C extends ElementType = "div"> =
  PolymorphicComponentPropsWithRef<C, ColumnsPropsBase>;

const ColumnComp = forwardRef(
  <C extends ElementType = "div">(
    { columns, dense, switchAt, as, gutter, ...props }: ColumnsProps<C>,
    ref?: PolymorphicRef<C>
  ) => {
    const maybePx = React.useMemo(() => {
      return typeof switchAt === "string"
        ? toPX(switchAt)
        : typeof switchAt === "number" && switchAt > -1
        ? switchAt
        : null;
    }, [switchAt]);

    const widthToSwitchAt: number = maybePx ?? 0; //zero is used to make the switchAt a noop

    const [shouldSwitch, safeRef] = useContainerQuery(
      { width: widthToSwitchAt },
      ref
    );

    return shouldSwitch ? (
      <Stack as={as} ref={safeRef} gutter={gutter} {...props} />
    ) : (
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      //@ts-expect-error
      <ColumnsBase
        as={as}
        ref={safeRef}
        gutter={gutter}
        columns={columns}
        dense={dense}
        {...props}
      />
    );
  }
);

export const Columns = styled(ColumnComp).attrs(({ as, forwardedAs }) => {
  return {
    forwardedAs: as ?? forwardedAs,
    as: ColumnComp,
  };
})``;

Columns.displayName = "Columns";

export interface ColumnProps {
  span?: number;
  offsetStart?: number;
  offsetEnd?: number;
}

const safeSpan = (span: unknown) => {
  return typeof span === "number" ? span : 1;
};

/**
 *
 * span is remaped to colSpan due to span being an attribute that gets
 * passed to the underlying element.  This can cause issues with Grid layout.
 *
 * In a future breaking change, colSpan should be the public API.
 * */
export const Column = styled.div.attrs<ColumnProps>(
  ({ span, style, offsetStart = 0, offsetEnd = 0 }) => {
    const safeOffsetStart = offsetStart > 0 ? offsetStart : undefined;
    const safeOffsetEnd = offsetEnd > 0 ? offsetEnd : undefined;
    return {
      "data-bedrock-column": "",
      span: undefined,
      style: {
        ...style,
        "--span": Math.max(safeSpan(span), 1),
        "--offsetStart": safeOffsetStart,
        "--offsetEnd": safeOffsetEnd,
      },
    };
  }
)<ColumnProps>`
  @property --span {
    syntax: "<number>";
    inherits: true;
    initial-value: 1;
  }

  @property --offsetStart {
    syntax: "<number>";
    inherits: true;
    initial-value: 1;
  }

  @property --offsetEnd {
    syntax: "<number>";
    inherits: true;
    initial-value: 1;
  }

  grid-column: span min(var(--span, 1), var(--columns, 1));

  &[style*="--offset"] {
    display: contents;
  }

  &[style*="--offset"] > * {
    grid-column: span min(var(--span, 1), var(--columns, 1));
  }

  &[style*="--offsetStart"]::before {
    content: "";
    grid-column: span min(var(--offsetStart, 1), var(--columns, 1));
  }

  &[style*="--offsetEnd"]::after {
    content: "";
    grid-column: span min(var(--offsetEnd, 1), var(--columns, 1));
  }
`;

Column.displayName = "Column";

/**
 * This module is adapted from https://github.com/mikolalysenko/to-px/blob/master/browser.js
 */

const PIXELS_PER_INCH = 96;

/* c8 ignore start */
function parseUnit(str: string): [number, string] {
  str = String(str);
  const num = parseFloat(str);

  /* c8 ignore next */
  const [, unit] = str.match(/[\d.\-+]*\s*(.*)/) ?? ["", ""];

  return [num, unit];
}

/* c8 ignore next */
function toPX(str: string, element?: Element): number | null {
  /* c8 ignore next */
  if (!str) return null;

  const elementOrBody = element ?? document.body;
  const safeStr = (str ?? "px").trim().toLowerCase();

  /* c8 ignore next */
  switch (safeStr) {
    case "vmin":
    case "vmax":
    case "vh":
    case "vw":
    case "%":
      return null;
    case "ch":
    case "ex":
      return getSizeBrutal(safeStr, elementOrBody);
    case "em":
      return getPropertyInPX(elementOrBody, "font-size");
    case "rem":
      return getPropertyInPX(document.body, "font-size");
    case "in":
      return PIXELS_PER_INCH;
    case "cm":
      return PIXELS_PER_INCH / 2.54;
    case "mm":
      return PIXELS_PER_INCH / 25.4;
    case "pt":
      return PIXELS_PER_INCH / 72;
    case "pc":
      return PIXELS_PER_INCH / 6;
    case "px":
      return 1;
    default: {
      const [value, units] = parseUnit(safeStr);

      if (isNaN(value)) return null;

      if (!units) return value;

      const px = toPX(units, element);
      return typeof px === "number" ? value * px : null;
    }
  }
}

/* c8 ignore next */
function getPropertyInPX(element: Element, prop: string): number {
  const [value, units] = parseUnit(
    getComputedStyle(element).getPropertyValue(prop)
  );
  return value * (toPX(units, element) ?? 1);
}

function getSizeBrutal(unit: string, element: Element) {
  const testDIV = document.createElement("div");
  testDIV.style["height"] = "128" + unit;
  element.appendChild(testDIV);
  const size = getPropertyInPX(testDIV, "height") / 128;
  element.removeChild(testDIV);
  return size;
}
/* c8 ignore stop */
