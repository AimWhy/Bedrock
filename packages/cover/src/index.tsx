import {
  CSSLength,
  Gutter,
  SizesOptions,
  getSafeGutter,
  getSizeValue,
  useTheme,
} from "@bedrock-layout/spacing-constants";
import { forwardRefWithAs } from "@bedrock-layout/type-utils";
import React, { CSSProperties } from "react";

/**
 * `minHeight` can be a CSSLength, a number, or a key of the theme's sizes object
 */
export type MinHeight = CSSLength | number | SizesOptions;

/**
 * Props for the Cover component.
 */
export type CoverProps = {
  /**
   * Slot to be placed at the top of the cover component, above the centered content.
   */
  top?: React.ReactNode;
  /**
   * Slot to be placed at the bottom of the cover component, below the centered content.
   */
  bottom?: React.ReactNode;
  /**
   * Sets space between each element.
   * @deprecated Use `gap` instead.
   */
  gutter?: Gutter;
  /**
   * Sets space between each element.
   */
  gap?: Gutter;
  /**
   * Sets the minimum height of the cover component.
   * `minHeight` can be a CSSLength, a number, or a key of the theme's sizes object
   * @default "100vh"
   */
  minHeight?: MinHeight;
  /**
   * Sets the content to stretch to the full height of the cover component minus the top and bottom slots.
   */
  stretchContent?: boolean;
};

/**
 * The `Cover` component is designed to vertically cover a predefined area, `100vh` by default, and vertically center its children.
 * You can also conditionally render a top and/or bottom slot as well.
 */
export const Cover = forwardRefWithAs<"div", CoverProps>(function Cover(
  {
    as: Component = "div",
    children,
    gap,
    gutter,
    top,
    bottom,
    minHeight,
    style = {},
    //TODO: Change from boolean to string union of "stretch-content" | "default"
    stretchContent,
    ...props
  },
  ref,
) {
  const theme = useTheme();
  const maybeGutter = getSafeGutter(theme, gap ?? gutter);
  const maybeMinHeight = getSizeValue(theme, minHeight);

  const attributeVal = stretchContent === true ? "stretch-content" : "";

  return (
    <Component
      ref={ref}
      data-bedrock-cover={attributeVal}
      style={
        {
          "--gutter": maybeGutter,
          "--minHeight": maybeMinHeight,
          ...style,
        } as CSSProperties
      }
      {...props}
    >
      {top && <div data-bedrock-cover-top="">{top}</div>}
      <div data-bedrock-cover-centered="">{children}</div>
      {bottom && <div data-bedrock-cover-bottom="">{bottom}</div>}
    </Component>
  );
});
