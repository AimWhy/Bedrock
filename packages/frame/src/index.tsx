import { forwardRefWithAs } from "@bedrock-layout/type-utils";
import React, { CSSProperties } from "react";
type RatioString = `${number}/${number}` | `${number} / ${number}`;

type Ratio = [number, number] | RatioString;
export interface FrameProps {
  ratio?: Ratio;
  position?: string;
}

function checkIsRatio(ratio: unknown): ratio is Ratio {
  const isCorrectArray =
    Array.isArray(ratio) && ratio.length === 2 && ratio.every(Number.isFinite);
  return (
    isCorrectArray ||
    (typeof ratio === "string" &&
      /^\d{1,1000} {0,1}\/ {0,1}\d{1,1000}$/.test(ratio))
  );
}

function getRatioString(ratio: Ratio): RatioString {
  return Array.isArray(ratio) ? (ratio.join("/") as RatioString) : ratio;
}

function getSafeRatio(ratio: unknown): RatioString | undefined {
  const isRatio = checkIsRatio(ratio);

  return isRatio ? getRatioString(ratio) : undefined;
}

export const Frame = forwardRefWithAs<"div", FrameProps>(
  ({ as, ratio, style, position, ...props }, ref) => {
    const safeRatio = getSafeRatio(ratio);
    const safeStyle = style ?? {};

    const Component = as ?? "div";

    return (
      <Component
        data-bedrock-frame
        ref={ref}
        style={
          {
            ...safeStyle,
            "--ratio": safeRatio,
            "--position": position,
          } as CSSProperties
        }
        {...props}
      />
    );
  }
);

Frame.displayName = "Frame";
