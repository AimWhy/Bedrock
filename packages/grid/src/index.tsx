import {
  SizesOptions,
  SpacingOptions,
  getSizeValue,
  getSpacingValue,
  sizes,
} from "@bedrock-layout/spacing-constants";
import PropTypes from "prop-types";
import styled from "styled-components";

type MinItemWidth = number | string | SizesOptions;
export interface GridProps {
  gutter: keyof SpacingOptions;
  minItemWidth?: MinItemWidth;
}

function getSafeMinItemWidth(minItemWidth?: MinItemWidth) {
  if (typeof minItemWidth === "string") {
    return minItemWidth;
  }

  return typeof minItemWidth === "number" ? `${minItemWidth}px` : sizes.small;
}

export const Grid = styled.div.attrs<GridProps>(() => {
  return {
    "data-bedrock-grid": "",
  };
})<GridProps>`
  box-sizing: border-box;
  > * {
    margin: 0;
  }

  @property --minItemWidth {
    syntax: "<length-percentage>";
    inherits: false;
    initial-value: ${sizes.small};
  }

  --gutter: ${({ gutter, theme }) => getSpacingValue(theme, gutter) ?? "0px"};
  --minItemWidth: ${({ minItemWidth, theme }) =>
    getSafeMinItemWidth(getSizeValue(theme, minItemWidth) ?? minItemWidth)};

  display: grid;
  gap: var(--gutter);

  grid-template-columns: repeat(
    auto-fit,
    minmax(min(${sizes.small}, 100%), 1fr)
  );

  @supports (
    width:
      ${({ minItemWidth, theme }) =>
        getSafeMinItemWidth(getSizeValue(theme, minItemWidth) ?? minItemWidth)}
  ) {
    grid-template-columns: repeat(
      auto-fit,
      minmax(min(var(--minItemWidth), 100%), 1fr)
    );
  }
`;

Grid.displayName = "Grid";

Grid.propTypes = {
  gutter: PropTypes.string.isRequired as React.Validator<keyof SpacingOptions>,
  minItemWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};
