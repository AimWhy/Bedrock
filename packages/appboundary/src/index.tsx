import { Center } from "@bedrock-layout/center";
import { mergeBreakpoints } from "@bedrock-layout/spacing-constants";
import React from "react";
import styled, { ThemeContext } from "styled-components";

const safeTheme = { breakPoints: {} };

export const AppBoundary = styled.div.attrs((props) => {
  const { breakPoints } = React.useContext(ThemeContext) ?? safeTheme;
  const maxWidth = mergeBreakpoints(breakPoints).xxlarge;
  return {
    "data-bedrock-layout-appboundary": "",
    children: <Center maxWidth={maxWidth}>{props.children}</Center>,
  };
})`
  padding: 0;
  max-width: 100%;
  overflow: hidden;
  height: 100%;
`;

AppBoundary.displayName = "AppBoundary";
