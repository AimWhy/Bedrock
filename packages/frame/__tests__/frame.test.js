import React from "react";
import { create } from "react-test-renderer";

import { Frame } from "../src";

describe("Frame", () => {
  describe("correct usage", () => {
    test("Frame is not null", () => {
      expect(Frame).toBeTruthy();
    });

    it("renders with ratio as an array", () => {
      const frame = create(
        <Frame ratio={[16, 9]}>
          <img src="https://picsum.photos/5000" alt="random thing" />
        </Frame>
      );
      expect(frame.toJSON()).toMatchSnapshot();
    });

    it("renders with ratio as a string", () => {
      const frame = create(
        <Frame ratio="16/9">
          <img src="https://picsum.photos/5000" alt="random thing" />
        </Frame>
      );
      expect(frame.toJSON()).toMatchSnapshot();
    });

    it("renders without ratio", () => {
      const frame = create(
        <Frame>
          <img src="https://picsum.photos/5000" alt="random thing" />
        </Frame>
      );
      expect(frame.toJSON()).toMatchSnapshot();
    });

    it("renders custom position", () => {
      const frame = create(
        <Frame ratio={[16, 9]} position="top left">
          <img src="https://picsum.photos/5000" alt="random thing" />
        </Frame>
      );
      expect(frame.toJSON()).toMatchSnapshot();
    });
  });

  describe("incorrect usage", () => {
    beforeEach(() => {
      jest.spyOn(console, "error");
      console.error.mockImplementation(() => undefined);
    });
    afterEach(() => {
      console.error.mockRestore();
    });

    it("renders default position with incorrect value", () => {
      expect(console.error).not.toBeCalled();

      const errorStack = create(
        <Frame ratio={[16, 9]} position={true}>
          <img src="https://picsum.photos/5000" alt="random thing" />
        </Frame>
      );

      expect(console.error).toBeCalled();
      expect(errorStack.toJSON()).toMatchSnapshot();
    });

    it("renders without ratio if ratio is not correct type", () => {
      expect(console.error).not.toBeCalled();

      const errorStack = create(
        <Frame ratio={{ 0: 16, 1: 9 }}>
          <img src="https://picsum.photos/5000" alt="random thing" />
        </Frame>
      );

      expect(console.error).toBeCalled();
      expect(errorStack.toJSON()).toMatchSnapshot();
    });

    it("renders without ratio with error if array of length <1 provided", () => {
      expect(console.error).not.toBeCalled();

      const errorStack = create(
        <Frame ratio={[16]}>
          <img src="https://picsum.photos/5000" alt="random thing" />
        </Frame>
      );

      expect(console.error).toBeCalled();
      expect(errorStack.toJSON()).toMatchSnapshot();
    });

    it("renders without ratio with error if array of length >2 provided", () => {
      expect(console.error).not.toBeCalled();

      const errorStack = create(
        <Frame ratio={[16, 9, 8, 6]}>
          <img src="https://picsum.photos/5000" alt="random thing" />
        </Frame>
      );

      expect(console.error).toBeCalled();
      expect(errorStack.toJSON()).toMatchSnapshot();
    });

    it("renders without ratio with error if array of not numbers provided", () => {
      expect(console.error).not.toBeCalled();

      const errorStack = create(
        <Frame ratio={["16", "9"]}>
          <img src="https://picsum.photos/5000" alt="random thing" />
        </Frame>
      );

      expect(console.error).toBeCalled();
      expect(errorStack.toJSON()).toMatchSnapshot();
    });

    it("renders without ratio with error if ratio string is not correct format", () => {
      expect(console.error).not.toBeCalled();

      const errorStack = create(
        <Frame ratio="16:9">
          <img src="https://picsum.photos/5000" alt="random thing" />
        </Frame>
      );

      expect(console.error).toBeCalled();
      expect(errorStack.toJSON()).toMatchSnapshot();
    });
  });
});
