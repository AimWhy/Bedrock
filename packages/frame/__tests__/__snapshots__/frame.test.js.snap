// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frame correct usage renders custom position 1`] = `
.c0 {
  --n: 16;
  --d: 9;
  box-sizing: border-box;
  display: block;
  inline-size: 100%;
  position: relative;
  aspect-ratio: var(--n) / var(--d);
}

.c0 > * {
  overflow: hidden;
  position: absolute;
  -webkit-inset-block-start: 0;
  -ms-intb-rlock-start: 0;
  inset-block-start: 0;
  -webkit-inset-block-end: 0;
  -ms-inlrock-end: 0;
  inset-block-end: 0;
  inset-inline-start: 0;
  inset-inline-end: 0;
  inset-block: 0;
  inset-inline: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > :is(img,video) {
  inline-size: 100%;
  -webkit-block-size: 100%;
  -ms-flex-block-size: 100%;
  block-size: 100%;
  size: 100%;
  object-fit: cover;
  object-position: top left;
}

@supports not (aspect-ratio:1/1) {
  .c0 {
    padding-block-end: calc(var(--d) / var(--n) * 100%);
  }
}

<div
  className="c0"
  data-bedrock-frame=""
>
  <img
    alt="random thing"
    src="https://picsum.photos/5000"
  />
</div>
`;

exports[`Frame correct usage renders with ratio 1`] = `
.c0 {
  --n: 16;
  --d: 9;
  box-sizing: border-box;
  display: block;
  inline-size: 100%;
  position: relative;
  aspect-ratio: var(--n) / var(--d);
}

.c0 > * {
  overflow: hidden;
  position: absolute;
  -webkit-inset-block-start: 0;
  -ms-intb-rlock-start: 0;
  inset-block-start: 0;
  -webkit-inset-block-end: 0;
  -ms-inlrock-end: 0;
  inset-block-end: 0;
  inset-inline-start: 0;
  inset-inline-end: 0;
  inset-block: 0;
  inset-inline: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > :is(img,video) {
  inline-size: 100%;
  -webkit-block-size: 100%;
  -ms-flex-block-size: 100%;
  block-size: 100%;
  size: 100%;
  object-fit: cover;
  object-position: 50%;
}

@supports not (aspect-ratio:1/1) {
  .c0 {
    padding-block-end: calc(var(--d) / var(--n) * 100%);
  }
}

<div
  className="c0"
  data-bedrock-frame=""
>
  <img
    alt="random thing"
    src="https://picsum.photos/5000"
  />
</div>
`;

exports[`Frame correct usage renders without ratio 1`] = `
.c0 {
  --n: 1;
  --d: 1;
  box-sizing: border-box;
  display: block;
  inline-size: 100%;
  position: relative;
}

.c0 > * {
  overflow: hidden;
  position: absolute;
  -webkit-inset-block-start: 0;
  -ms-intb-rlock-start: 0;
  inset-block-start: 0;
  -webkit-inset-block-end: 0;
  -ms-inlrock-end: 0;
  inset-block-end: 0;
  inset-inline-start: 0;
  inset-inline-end: 0;
  inset-block: 0;
  inset-inline: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > :is(img,video) {
  inline-size: 100%;
  -webkit-block-size: 100%;
  -ms-flex-block-size: 100%;
  block-size: 100%;
  size: 100%;
  object-fit: cover;
  object-position: 50%;
}

<div
  className="c0"
  data-bedrock-frame=""
>
  <img
    alt="random thing"
    src="https://picsum.photos/5000"
  />
</div>
`;

exports[`Frame incorrect usage falls back to 1 with error if array of length <1 provided 1`] = `
.c0 {
  --n: 16;
  --d: 1;
  box-sizing: border-box;
  display: block;
  inline-size: 100%;
  position: relative;
  aspect-ratio: var(--n) / var(--d);
}

.c0 > * {
  overflow: hidden;
  position: absolute;
  -webkit-inset-block-start: 0;
  -ms-intb-rlock-start: 0;
  inset-block-start: 0;
  -webkit-inset-block-end: 0;
  -ms-inlrock-end: 0;
  inset-block-end: 0;
  inset-inline-start: 0;
  inset-inline-end: 0;
  inset-block: 0;
  inset-inline: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > :is(img,video) {
  inline-size: 100%;
  -webkit-block-size: 100%;
  -ms-flex-block-size: 100%;
  block-size: 100%;
  size: 100%;
  object-fit: cover;
  object-position: 50%;
}

@supports not (aspect-ratio:1/1) {
  .c0 {
    padding-block-end: calc(var(--d) / var(--n) * 100%);
  }
}

<div
  className="c0"
  data-bedrock-frame=""
>
  <img
    alt="random thing"
    src="https://picsum.photos/5000"
  />
</div>
`;

exports[`Frame incorrect usage falls back to 1 with error if array of length >2 provided 1`] = `
.c0 {
  --n: 16;
  --d: 9;
  box-sizing: border-box;
  display: block;
  inline-size: 100%;
  position: relative;
  aspect-ratio: var(--n) / var(--d);
}

.c0 > * {
  overflow: hidden;
  position: absolute;
  -webkit-inset-block-start: 0;
  -ms-intb-rlock-start: 0;
  inset-block-start: 0;
  -webkit-inset-block-end: 0;
  -ms-inlrock-end: 0;
  inset-block-end: 0;
  inset-inline-start: 0;
  inset-inline-end: 0;
  inset-block: 0;
  inset-inline: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > :is(img,video) {
  inline-size: 100%;
  -webkit-block-size: 100%;
  -ms-flex-block-size: 100%;
  block-size: 100%;
  size: 100%;
  object-fit: cover;
  object-position: 50%;
}

@supports not (aspect-ratio:1/1) {
  .c0 {
    padding-block-end: calc(var(--d) / var(--n) * 100%);
  }
}

<div
  className="c0"
  data-bedrock-frame=""
>
  <img
    alt="random thing"
    src="https://picsum.photos/5000"
  />
</div>
`;

exports[`Frame incorrect usage falls back to 1 with error if array of not numbers provided 1`] = `
.c0 {
  --n: 1;
  --d: 1;
  box-sizing: border-box;
  display: block;
  inline-size: 100%;
  position: relative;
  aspect-ratio: var(--n) / var(--d);
}

.c0 > * {
  overflow: hidden;
  position: absolute;
  -webkit-inset-block-start: 0;
  -ms-intb-rlock-start: 0;
  inset-block-start: 0;
  -webkit-inset-block-end: 0;
  -ms-inlrock-end: 0;
  inset-block-end: 0;
  inset-inline-start: 0;
  inset-inline-end: 0;
  inset-block: 0;
  inset-inline: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > :is(img,video) {
  inline-size: 100%;
  -webkit-block-size: 100%;
  -ms-flex-block-size: 100%;
  block-size: 100%;
  size: 100%;
  object-fit: cover;
  object-position: 50%;
}

@supports not (aspect-ratio:1/1) {
  .c0 {
    padding-block-end: calc(var(--d) / var(--n) * 100%);
  }
}

<div
  className="c0"
  data-bedrock-frame=""
>
  <img
    alt="random thing"
    src="https://picsum.photos/5000"
  />
</div>
`;

exports[`Frame incorrect usage renders default position with incorrect value 1`] = `
.c0 {
  --n: 16;
  --d: 9;
  box-sizing: border-box;
  display: block;
  inline-size: 100%;
  position: relative;
  aspect-ratio: var(--n) / var(--d);
}

.c0 > * {
  overflow: hidden;
  position: absolute;
  -webkit-inset-block-start: 0;
  -ms-intb-rlock-start: 0;
  inset-block-start: 0;
  -webkit-inset-block-end: 0;
  -ms-inlrock-end: 0;
  inset-block-end: 0;
  inset-inline-start: 0;
  inset-inline-end: 0;
  inset-block: 0;
  inset-inline: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > :is(img,video) {
  inline-size: 100%;
  -webkit-block-size: 100%;
  -ms-flex-block-size: 100%;
  block-size: 100%;
  size: 100%;
  object-fit: cover;
  object-position: 50%;
}

@supports not (aspect-ratio:1/1) {
  .c0 {
    padding-block-end: calc(var(--d) / var(--n) * 100%);
  }
}

<div
  className="c0"
  data-bedrock-frame=""
>
  <img
    alt="random thing"
    src="https://picsum.photos/5000"
  />
</div>
`;

exports[`Frame incorrect usage renders ratio of 1:1 with error if ratio is not an array 1`] = `
.c0 {
  --n: 16;
  --d: 9;
  box-sizing: border-box;
  display: block;
  inline-size: 100%;
  position: relative;
  aspect-ratio: var(--n) / var(--d);
}

.c0 > * {
  overflow: hidden;
  position: absolute;
  -webkit-inset-block-start: 0;
  -ms-intb-rlock-start: 0;
  inset-block-start: 0;
  -webkit-inset-block-end: 0;
  -ms-inlrock-end: 0;
  inset-block-end: 0;
  inset-inline-start: 0;
  inset-inline-end: 0;
  inset-block: 0;
  inset-inline: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > :is(img,video) {
  inline-size: 100%;
  -webkit-block-size: 100%;
  -ms-flex-block-size: 100%;
  block-size: 100%;
  size: 100%;
  object-fit: cover;
  object-position: 50%;
}

@supports not (aspect-ratio:1/1) {
  .c0 {
    padding-block-end: calc(var(--d) / var(--n) * 100%);
  }
}

<div
  className="c0"
  data-bedrock-frame=""
>
  <img
    alt="random thing"
    src="https://picsum.photos/5000"
  />
</div>
`;
